crates/ruff_python_formatter/src/expression/string.rs

```diff
@@ -639,3 +655,301 @@ fn normalize_string(
+/// we see indentation that contains a tab or any other none ascii-space whitespace we rewrite the
+/// string.
+///
+/// Additionally, if any line in the docstring has less indentation than the docstring
+/// (effectively a negative indentation wrt. to the current level), we pad all lines to the
+/// level of the docstring with spaces.
+/// ```python
+/// def f():
+///    """first line
+/// line a
+///    line b
+///      line c
+///    """
+/// ```
+/// Here line a is 3 columns negatively indented, so we pad all lines by an extra 3 spaces:
+/// ```python
+/// def f():
+///    """first line
+///    line a
+///       line b
+///         line c
+///    """
+/// ```
+fn format_docstring(
+    // The start of the string prefix to after the closing quotes
+    outer_range: TextRange,
+    f: &mut PyFormatter,
+) -> FormatResult<()> {
+    let locator = f.context().locator();
+    let string_part = FormatStringPart::new(
+        outer_range,
+        // It's not an f-string
+        Quoting::CanChange,
+        &locator,
+        f.options().quote_style(),
+    );
+
+    // Black doesn't change the indentation of docstrings that contain an escaped newline
+    if locator.slice(outer_range).contains("\\\n") {
+        return string_part.fmt(f);
+    }
+
+    let (normalized, _) = normalize_string(
+        locator.slice(string_part.range),
+        string_part.preferred_quotes,
+        string_part.is_raw_string,
+    );
+    // is_borrowed is unstable :/
+    let already_normalized = matches!(normalized, Cow::Borrowed(_));
+
+    let mut lines = normalized.lines().peekable();
+
+    // Start the string
+    write!(f, [string_part.prefix, string_part.preferred_quotes])?;
+    // We track where in the source docstring we are (in source code byte offsets)
+    let mut offset = string_part.range.start();
+
+    // The first line directly after the opening quotes has different rules than the rest, mainly
+    // that we remove all leading whitespace as there's no indentation
+    let first = lines.next().unwrap_or_default();
+    let trim_end = first.trim_end();
+
+    // Edge case: The first line is `""" "content`, so we need to insert chaperone space to
+    // avoid `""""content`
+    if trim_end
+        .trim_start()
+        .starts_with(string_part.preferred_quotes.style.as_char())
+    {
+        space().fmt(f)?;
+    }
+
+    if !trim_end.is_empty() {
+        // For the first line of the docstring we strip the leading and trailing whitespace, e.g.
+        // `"""   content   ` to `"""content`
+        let leading_whitespace = trim_end.text_len() - trim_end.trim_start().text_len();
+        let trimmed_line_range =
+            TextRange::at(offset, trim_end.text_len()).add_start(leading_whitespace);
+        if already_normalized {
+            source_text_slice(trimmed_line_range, ContainsNewlines::No).fmt(f)?;
+        } else {
+            dynamic_text(trim_end.trim_start(), Some(trimmed_line_range.start())).fmt(f)?;
+        }
```

Comment:
You want to use `trim_whitespace_start()` and `trim_whitespace_end()` to only trim valid python whitespace.

Reply:
i tried this, but to my surprise black disagrees and trims like rust instead. This is the diffs i get i make that change: https://gist.github.com/konstin/62227a242cea073a0d62efc19448e4fb